moq-pub/src/media_runner.rs

```diff
@@ -0,0 +1,138 @@
+use crate::media::{self, MapSource};
+use log::debug;
+use moq_transport::message::Message;
+use moq_transport::message::{Announce, SubscribeError};
+use moq_transport::{object, Object, VarInt};
+use std::collections::HashMap;
+use std::sync::Arc;
+use tokio::io::AsyncWriteExt;
+use tokio::sync::broadcast;
+use tokio::sync::mpsc;
+use tokio::task::JoinSet;
+
+use webtransport_generic::Session as WTSession;
+
+pub struct MediaRunner<S: WTSession> {
+	send_objects: object::Sender<S>,
+	outgoing_ctl_sender: mpsc::Sender<Message>,
+	incoming_ctl_receiver: broadcast::Receiver<Message>,
+	source: Arc<MapSource>,
+}
+
+impl<S: WTSession> MediaRunner<S> {
+	pub async fn new(
+		send_objects: object::Sender<S>,
+		outgoing: mpsc::Sender<Message>,
+		incoming: (broadcast::Receiver<Message>, broadcast::Receiver<Object>),
+	) -> anyhow::Result<Self> {
+		let outgoing_ctl_sender = outgoing;
+		let (incoming_ctl_receiver, _incoming_obj_receiver) = incoming;
+		Ok(Self {
+			send_objects,
+			outgoing_ctl_sender,
+			incoming_ctl_receiver,
+			source: Arc::new(MapSource::default()),
+		})
+	}
+	pub async fn announce(&mut self, namespace: &str, source: Arc<media::MapSource>) -> anyhow::Result<()> {
+		debug!("media_runner.announce()");
+		// Only allow one souce at a time for now?
+		self.source = source;
+
+		// ANNOUNCE the namespace
+		self.outgoing_ctl_sender
+			.send(Message::Announce(Announce {
+				track_namespace: namespace.to_string(),
+			}))
+			.await?;
+
+		// wait for the go ahead
+		loop {
+			if let Message::AnnounceOk(_) = self.incoming_ctl_receiver.recv().await? {
```

Comment:
Unfortunately this will race with any other incoming messages. I don't think it will break currently but it's going to a blocker eventually.

Reply:
Also what about `AnnounceError`?

Reply:
I do need to handle the error case here, yes, but the race shouldn't be an issue because we just loop if we get something other that what we're waiting for. `incoming_ctl_receiver` is a broadcast receiver so that if code elsewhere wants to handle a different message, it should still be able to get it.

Reply:
Re-reading this part of the code to add error handling, you might still be right about control message races here... `incoming_ctl_receiver` _is_ a broadcast receiver (essentially the inverse of an `mpsc` channel), **but** some of the other receivers are created by resubscribing to **this same instance** _after_ it would have potentially already consumed unrelated messages. Leaving a TODO to revisit that later. :-/